Implement Tracker Commands

debbiextan opened the pull request:
Tracker Commands include:
- StartTrackerCommand
- PauseTrackerCommand
- StopTrackerCommand
- ViewTrackerCommand

Tracker-related classes include:
- Tracker
- TrackerList

Tracker-related file: tracker.txt

debbiextan committed:
Implement PauseTrackerCommand.
SLAPped StartTrackerCommand and StopTrackerCommand.

debbiextan committed:
Fixed Warnings related to access modifiers in Ui class.

debbiextan committed:
Allow Tracker-related Commands to detect Task descriptions with whitespaces.

debbiextan committed:
Merge branch 'master' into master

debbiextan committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	output.txt
#	src/main/java/oof/CommandParser.java
#	src/main/java/oof/Ui.java
#	src/main/java/oof/command/StartTrackerCommand.java
#	src/main/java/oof/command/StopTrackerCommand.java

debbiextan committed:
Implement PauseTrackerCommand.

debbiextan committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/oof/Ui.java
#	src/main/java/oof/command/StartTrackerCommand.java
#	src/main/java/oof/command/StopTrackerCommand.java

debbiextan committed:
Merge remote-tracking branch 'upstream/master'

debbiextan committed:
Update User Guide to display Task Tracker related Commands.
Refactored Help, Summary and Tracker Commands to include sample run screenshots.

debbiextan committed:
Refactor SubHeading for Tracker Commands.
Edit tone to be more casual.

debbiextan committed:
Refactor SubHeading for Tracker Commands.

debbiextan committed:
Remove unnecessary JavaDocs for execute and isExit methods in StartTrackerCommand and PauseTrackerCommand classes.
Add template for viewTracker-related segments in User Guide and Developer Guide.

debbiextan committed:
Implement ViewTracker.
SLAP Summary Command.
Edit OofException message for PauseTracker and StopTracker Command.

debbiextan committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	output.txt

debbiextan committed:
Refactor implementation of Tracker.
Add Storage function for TrackerList.
Add TrackerList parameter to execute method.

debbiextan committed:
Fix Logic errors in Start, Stop and Pause Tracker Commands.
Update UG with new syntax for Tracker related Commands.

debbiextan committed:
Include tracker.txt

jasperosy reviewed (commented):
DeveloperGuide:
I have already included the capability to automatically assign figure caption number for each diagram, please adhere to the format and apply an anchor and figure label to the diagrams you have added.
Please don't submit an implementation that is half done. Consider a top-down approach by thinking of possible alternatives before you start coding. In that case, your rationale and alternatives considered section would not be blank.

UserGuide:
Try not to just create subfolders to store your images as I am not sure if the bot will pick them up. 
Remember to include an anchor and figure label for each diagram.
Please put in some effort to at least state an example usage of your feature. e.g when a user enters the command.... this happens...

Ui:
```suggestion
        for (int i = 0; i < trackerList.getSize(); i++) {
            oof.model.tracker.Tracker tracker = trackerList.getTracker(i);
            int timeTaken = (int) tracker.getTimeTaken();
            timeTaken /= 10;
```
Magic number in method. Even if its just division, please insert a meaningful const. 
There is a dirty typecast to tracker.getTimeTaken(), please rectify.

Output.txt:
Why is there an entry with a /add description?
```Suggestion
T||N||/add buy chocolates||28-10-2019||||||
```
Is that just for testing or are you using doing some string parsing?

ViewTrackerCommand:
```Suggestion
    public void execute(SemesterList semesterList, TaskList tasks, TrackerList trackerList, Ui ui, Storage storage)
            throws OofException {
        trackerList = timeSpentByModule(tasks);
        if (trackerList.getSize() == 0) {
            throw new OofException("There are no completed Tasks!");
        }
```
Magic number, try using isEmpty() or set a constant

Storage:
```Suggestion
    private Tracker processLine(String line) {
        String[] processed = line.split("\t");
        String module = processed[0];
        String startDate = processed[1];
        String lastUpdated = processed[2];
        long timeTaken = Long.parseLong(processed[3]);
        return new Tracker(module, startDate, lastUpdated, timeTaken);
    }
```
More magic numbers, please rectify

kahlokkee commented on docs/UserGuide.adoc:
```diff
@@ -360,6 +366,64 @@ Opens menu to view, add, edit and remove Semester and Module data.
+Example: `viewTracker Day`
+
+image::demo_images/Demo_ViewTrackerTimePeriod.png[]
+
```
Why is the task tracker tracking module when it is a TASK tracker and module is not a TASK. This feature is meant to track individual tasks such that the user can know how much time they spent on each task. Tracking time spent on a module does not provide any insightful details to the user. There is a reason why each module contains a list of assignments. "Time spent" belongs to the task (assignment) and when you navigate to a module, you can navigate to each individual assignment and view the time taken for that assignment.

Also, there should not be 4 different commands for the same implementation as it will cause confusion for the user and it does not fit with the command style that we use.

Lastly, the UG is missing the source files for your images. Is it intentional for the UG to be merged while incomplete?

debbiextan replied:
I have changed the implementation of the Tracker feature as suggested. UG has also been corrected.

kahlokkee commented on src/main/java/oof/CommandParser.java:
```diff
@@ -95,8 +97,14 @@ public static Command parse(String line) throws OofException {
             return new SemesterCommand();
+        case "viewTracker":
+            line = line.replaceFirst("viewTracker", "").trim();
+            return new ViewTrackerCommand(line);
```
Refrain from using camelcase for your commands as it is inconsistent with our past practices

debbiextan replied:
This issue has been rectified as suggested.

kahlokkee commented on src/main/java/oof/Storage.java:
```diff
@@ -161,6 +165,61 @@ public void writeTaskList(TaskList taskList) {
+            File file = new File(PATH_TRACKER);
+            FileReader fileReader = new FileReader(file);
+            BufferedReader br = new BufferedReader(fileReader);
+            ArrayList<oof.model.tracker.Tracker> trackerList = new ArrayList<>();
```
do not use such conventions.

debbiextan replied:
This issue has been rectified as suggested.

kahlokkee commented on src/main/java/oof/Ui.java:
```diff
@@ -733,7 +734,7 @@ public void printDeadlineAndEvent(String taskTime, String taskName) {
+    private void printTodo(String taskName) {
         if (taskName.length() > TEXT_SIZE_LONG) {
             taskName = taskName.substring(DESCRIPTION_LONG_START, DESCRIPTION_LONG_END);
             System.out.print(" " + taskName + ".. |");
```
Why are these even changed? Revert these changes.

debbiextan replied:
Changes have now been reverted.

kahlokkee commented on src/main/java/oof/command/Command.java:
```diff
@@ -28,12 +29,14 @@
      */
-    public abstract void execute(SemesterList semesterList, TaskList tasks, Ui ui, Storage storage) throws OofException;
+    public abstract void execute(SemesterList semesterList, TaskList tasks, TrackerList trackerList,
+                                 Ui ui, Storage storage) throws OofException;
```
Why is TrackerList added to the argument parameter when it is not used by most commands.

debbiextan replied:
This has now been rectified.

kahlokkee reviewed (changes requested):
Stop going around auto optimizing everything that you see.

debbiextan committed:
Move demo images for Help, IndividualHelp and Summary Commands to images file.
Refactor viewTracker Command format in CommandParser as suggested.
Revert Ui methods access modifiers.
Fix declaration conventions for Tracker objects.
Remove TrackerList variable from Command.execute method.
Refactor Storage.readTrackerList.
Fix magic numbers in ScheduleCommand.
Re-Implement Start, Stop, Pause and ViewTracker Commands.
Add relevant UG images to Start, Stop, Pause and ViewTracker Commands.

debbiextan committed:
Fix Travis.

debbiextan committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	docs/UserGuide.adoc
#	output.txt
#	src/main/java/oof/CommandParser.java
#	src/main/java/oof/Storage.java
#	src/main/java/oof/command/SemesterCommand.java

debbiextan committed:
Fix merge conflicts.

debbiextan committed:
Change tone of UG to be more friendly.
Fixed merge conflicts.

debbiextan commented:
Code changes for v1.3 are done.
UG tone edited to be more friendly.

Do take a look.

debbiextan committed:
Reorganise User Guide by Command types.
Update help manual.txt.

debbiextan committed:
Add Semester related Command.
Fix TOC level and sect num to take more sections.

kahlokkee commented on src/main/java/oof/command/StartTrackerCommand.java:
```diff
@@ -2,88 +2,113 @@
+                storage.writeTrackerList(trackerList);
+                ui.printStartAtCurrent(tracker);
+            }
         }
```
This method of implementation does not take into account the possibility of multiple assignments having the same name (e.g. "assignment 1" for different modules). If there are multiple assignments with the same name, some of them are unable to be tracked at all.

debbiextan replied:
The code has been rectified as suggested.
Start, Stop and Pause commands now take in both module code and description as parameters.

kahlokkee commented on src/main/java/oof/model/tracker/Tracker.java:
```diff
@@ -0,0 +1,103 @@
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class Tracker extends Assignment {
```
Why is tracker inheriting from assignment when it is not a "task" per se.

debbiextan replied:
The code has been rectified as suggested.
Tracker now makes reference to Assignment Tasks when necessary instead.

kahlokkee commented on src/main/java/oof/Storage.java:
```diff
@@ -177,6 +192,78 @@ public void writeTaskList(TaskList taskList) {
+        Date updated = readFormat.parse(lastUpdated);
+
+        return new Tracker(moduleCode, description, deadline, start, updated, timeTaken);
+    }
```
Status of the tracker is not being processed so it is always set to not completed even if it is saved as completed

debbiextan replied:
The code has been rectified as suggested.
The tracker now updates the TaskList's Task variable (Status property) correctly.

kahlokkee commented on src/main/java/oof/command/StopTrackerCommand.java:
```diff
@@ -26,52 +25,48 @@
+        Tracker tracker = trackerList.findTrackerByDesc(description);
+        boolean isCompleted = tracker.getStatus();
+
+        if (isCompleted) {
+            throw new OofException("Assignment has already been completed.");
```
This condition is never true because of your Storage bugs.

debbiextan replied:
The code is now refactored.

kahlokkee commented on src/main/java/oof/command/ViewTrackerCommand.java:
```diff
@@ -0,0 +1,114 @@
+            }
+        }
+        return moduleTrackerList;
+    }
```
Arrow head.

debbiextan replied:
The code has been rectified and now has methods where appropriate.

kahlokkee reviewed (changes requested):
Expect more rigorous testing especially for edge cases. I can think of multiple scenarios where your entire tracker will break permanently. The implementation of Tracker and its related classes are also a little weird. Task tracking is supposed to be a feature extension for Assignments but your implementation is totally independent of Assignment. For example, marking an Assignment as done using DoneCommand does not interact with the Tracker class at all and the Tracker are unable to see whether the Assignment is marked as done (Why does tracker even have its own "done" status??).

You'll have to seriously reconsider the feasibility and practicality of your implementation and how it should be integrated to the existing features rather than creating a totally independent feature.

debbiextan committed:
Update DG with ViewTrackerCommand.

debbiextan committed:
Update DG with ViewTrackerCommand.
Perform grammar check on DG.
Perform grammar check on UG.

debbiextan committed:
Refactor Start, Stop, Pause and View Tracker Command.
Update UG and DG with new Start, Stop and Pause Tracker Command syntax.
Update demo images for Start, Stop, Pause and View Tracker Command.

jasperosy reviewed (commented):
Manual.txt:
Go update based on UG. useless if its not updated for PE

Ui:
``` Suggestion
    public void printTrackerDiagram(ModuleTrackerList moduleTrackerList) {
        printLine();
        for (int i = 0; i < moduleTrackerList.getSize(); i++) {
            ModuleTracker moduleTracker = moduleTrackerList.getModuleTracker(i);
            int timeTaken = (int) moduleTracker.getTotalTimeTaken();
            int segmentedTimeTaken = timeTaken / TEN_MINUTES_BLOCK;
            for (int j = 0; j < segmentedTimeTaken; j++) {
                if (j == FIRST_VAR) {
                    System.out.println("| ");
                    System.out.print("| #");
                } else {
                    System.out.print("#");
                }
            }
            String moduleCode = moduleTracker.getModuleCode();
            System.out.print("\t" + moduleCode + " -- " + timeTaken + " minutes\n");
        }
    }
```
Nested loop, not enough abstraction. can do better at least for inner for loop.

StartTrackerCommand:
``` Suggestion
    public void execute(SemesterList semesterList, TaskList tasks, Ui ui, Storage storage) throws OofException {
        if (description.isEmpty()) {
            throw new OofException("Please enter a Task!");
            throw new OofException("Please enter the Assignment module code and description!");
        }
        Task task = findTask(tasks, description);
        if (task.getStatus()) {
            throw new OofException("Task has already been completed.");

        String[] input = description.split(" ", 2);
        if (input.length < MINIMUM_SIZE) {
            throw new OofException("Invalid input!");
        }
```
whats the 2. same in stoptrackercommand.

debbiextan committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	docs/UserGuide.adoc
#	output.txt

debbiextan committed:
Update manual.txt
Update UG.
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	docs/UserGuide.adoc
#	output.txt

debbiextan committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	docs/DeveloperGuide.adoc

debbiextan committed:
Refactor Start, Stop, Pause and View Tracker Command.
Refactor ViewTracker related methods in Ui.
Update manual.txt
Update UG.
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	docs/UserGuide.adoc
#	output.txt

debbiextan commented:
Thanks, @KahLokKee and @jasperosy for your reviews and suggestions. The code has been refactored alongside the UG and DG updated accordingly.

jasperosy reviewed (commented):
UserGuide:
mentioned in previous comments that images need to be labelled. Follow the labelling convention. 
Also, can you follow the tone of the previous features? You are not selling your feature but explaining your feature with you language. 
Follow example format, language, label images
Please follow format for consistency.

Manual:
i dont see any correct changes in your manual... Can you take a look at the command summary again and edit your manual accordingly
e.g. no sort command, recurring wrong, viewweek wrong, calendar wrong, reminder also wrong, missing out commands....

CommandParser:
Also, why are you tampering with my methods for CommandParser?? It was originally correct and now you deleted and pasted it back.

debbiextan committed:
Update manual.txt with available commands.
Revert CommandParser to original.
Update UG with standardised label images.

debbiextan commented:
@jasperosy 
The CommandParser edits were created unintentionally during a merge conflict fix. I've replaced all the methods manually from the latest master copy of CommandParser. Do take a look and double-check that your methods back to their original.

UG label images have been edited to fit our standardized format and edited tone for tracker related features. If tone of Semester features need editing, @Kenlhc will need to make his necessary edits as well.

Thanks for pointing out the manual.txt issue, it has been rectified according to the available commands we have noted down in the UG.

jasperosy commented:
> @jasperosy
> The CommandParser edits were created unintentionally during a merge conflict fix. I've replaced all the methods manually from the latest master copy of CommandParser. Do take a look and double-check that your methods back to their original.
> 
> UG label images have been edited to fit our standardized format and edited tone for tracker related features. If tone of Semester features need editing, @Kenlhc will need to make his necessary edits as well.
> 
> Thanks for pointing out the manual.txt issue, it has been rectified according to the available commands we have noted down in the UG.

That brings about another point. Have you ensured that your UG is up to date with the copy in the repo now? Because I have already rectified @Kenlhc part.
